### ISSUE TITLE
Problems with QoS > 0

### ISSUE THREAD

**Issue by swIsoftSrl** [chkr1011/MQTTnet#1089]
### Describe your question
Hi, i recently created 2 c# application with the MqttNet nuget. All works perfectly, but I have some problems with the redelivery of the QoS 1 and 2 messages not delivered (client offline).

My client connection code is like this:

`static async void StartClient()
        {
            //private System.Diagnostics.EventLog Log;

            //IsoftMQTTClient settings = new IsoftMQTTClient();
            //settings.LoadSettings();
            //StartServer();
            EventLog Log1 = new EventLog();

            if (!EventLog.SourceExists("IsMQTTClient"))
                EventLog.CreateEventSource("IsMQTTClient", "IsMQTTClient");

            Log1.Source = "IsMQTTClient";
            Log1.Log = "IsMQTTClient";
            

            try {
                // Create a new MQTT client.
                var factory = new MqttFactory();
                var mqttClient = factory.CreateMqttClient();

                MqttClientOptions options;

                //Log1.WriteEntry("aaa " + GlobalVar.MQTT_CLIENT_CLEAN_SESSION.ToUpper());

                if (GlobalVar.MQTT_CLIENT_CLEAN_SESSION.ToUpper() == "S")
                {
                    // Create TCP based options using the builder.
                    options = (MqttClientOptions)new MqttClientOptionsBuilder()
                        .WithClientId(GlobalVar.MQTT_CLIENT_CLIENT_ID_WINDOWS_SERVICE)
                        .WithTcpServer(GlobalVar.MQTT_BROKER_HOST, Int32.Parse((GlobalVar.MQTT_BROKER_PORT))) // Port is optional
                        .WithCredentials(GlobalVar.MQTT_BROKER_USERNAME, GlobalVar.MQTT_BROKER_PASSWORD)
                        .WithCommunicationTimeout(new TimeSpan(0, 1, 0))
                        //.WithTls()
                        .WithCleanSession()
                        .Build();
                }
                else
                {
                    options = (MqttClientOptions)new MqttClientOptionsBuilder()
                        .WithClientId(GlobalVar.MQTT_CLIENT_CLIENT_ID_WINDOWS_SERVICE)
                        .WithTcpServer(GlobalVar.MQTT_BROKER_HOST, Int32.Parse((GlobalVar.MQTT_BROKER_PORT))) // Port is optional
                        .WithCredentials(GlobalVar.MQTT_BROKER_USERNAME, GlobalVar.MQTT_BROKER_PASSWORD)
                        .WithCommunicationTimeout(new TimeSpan(0, 1, 0))
                        .WithCleanSession(false)
                        //.WithTls()
                        .Build();
                }

                //Log1.WriteEntry(options.CleanSession.ToString());

                await mqttClient.ConnectAsync(options, CancellationToken.None); // Since 3.0.5 with CancellationToken
                Log1.WriteEntry("### CONNECTED TO BROKER ###");
                //await mqttClient.SubscribeAsync(new MqttTopicFilterBuilder().WithTopic("#").Build());
                //Log1.WriteEntry("### SUBSCRIBED TO ALL TOPICS ###");
                SubscribeToTopics(GlobalVar.MQTT_CLIENT_TOPIC_SUBSCRIPTION, mqttClient, options);
                //while (true)
                //{
                await PoolMessagesAsync(mqttClient, options);
                //}

                //Console.ReadKey();
            } catch (Exception ex) { Log1.WriteEntry("Errore: " + ex.ToString()); }
            }`

inside SubscribeToTopics() i'm subscribing to all topics available, and while client is connected i recive every message the broker recives
`await mqttClient.SubscribeAsync(new MqttTopicFilterBuilder().WithTopic("#").Build());
            Log1.WriteEntry("### SUBSCRIBED TO ALL TOPICS ###");`

My Broker is hosted like this:
`static async void StartServer()
        {
            var factory = new MqttFactory();

            // Configure MQTT server.
            var optionsBuilder = new MqttServerOptionsBuilder()
                .WithConnectionValidator(c =>
                {
                    if (c.ClientId.Length < 1)
                    {
                        c.ReasonCode = MqttConnectReasonCode.ClientIdentifierNotValid;
                        return;
                    }

                    System.Diagnostics.EventLog Log1;

                    Log1 = new EventLog();

                    if (!EventLog.SourceExists("IsMQTTBroker"))
                        EventLog.CreateEventSource("IsMQTTBroker", "IsMQTTBroker");

                    Log1.Source = "IsMQTTBroker";
                    Log1.Log = "IsMQTTBroker";

                    //Log1.WriteEntry(GlobalVar.MQTT_BROKER_HOST);
                    //Log1.WriteEntry(GlobalVar.MQTT_BROKER_PORT);
                    //Log1.WriteEntry(GlobalVar.MQTT_BROKER_USERNAME);
                    //Log1.WriteEntry(GlobalVar.MQTT_BROKER_PASSWORD);
                    //Log1.WriteEntry(GlobalVar.MQTT_CLIENT_CLIENT_ID_WINDOWS_SERVICE);
                    //
                    //Log1.WriteEntry("c Username " +c.Username);
                    //Log1.WriteEntry("c Password " + c.Password);
                    //Log1.WriteEntry("Int32.Parse((GlobalVar.MQTT_BROKER_PORT)) " + Int32.Parse((GlobalVar.MQTT_BROKER_PORT)).ToString());

                    if (c.Username != GlobalVar.MQTT_BROKER_USERNAME)
                    {
                        c.ReasonCode = MqttConnectReasonCode.BadUserNameOrPassword;
                        return;
                    }

                    if (c.Password != GlobalVar.MQTT_BROKER_PASSWORD)
                    {
                        c.ReasonCode = MqttConnectReasonCode.BadUserNameOrPassword;
                        return;
                    }

                    c.ReasonCode = MqttConnectReasonCode.Success;
                })
                .WithConnectionBacklog(100)
                .WithDefaultEndpointPort(Int32.Parse((GlobalVar.MQTT_BROKER_PORT)))
                .WithPersistentSessions();

            var mqttServer = new MqttFactory().CreateMqttServer();
            await mqttServer.StartAsync(optionsBuilder.Build());

            Console.WriteLine("Press any key to exit.");
            Console.ReadLine();
            //await mqttServer.StopAsync();
        }`

when i stop my client, send some messages with QoS 1 or 2 and start my client again those messages are vanished.
I have a static and unique ClientID, and the client is always started with ".WithCleanSession(false)". Any idea on what i'm missing to manage those messages recived while offline?

Thanks

Best regards,
Andrea

### Which project is your question related to?
- Client
- Server

**Comment by davemcdermid** [comment]
I had this same issue, have just figured it out. Update your SubscribeAsync call to set QoS:

    await mqttClient.SubscribeAsync(
      new MqttTopicFilterBuilder()
        .WithTopic("#")
        .WithAtLeastOnceQoS() // <-- This sets the QoS of the subscription
        .Build());

Hope this helps

**Comment by swIsoftSrl** [comment]
Hi, i had already tested in the past days to set this, but with no luck. Thanks anyway for the tip.
I tested with your exact code and the client didn't get the messages sent while it was offline. I tested both QoS in my subscribe async method.

now my subscription handler looks like this
`static async void SubscribeToTopics(string modalita, IMqttClient mqttClient, IMqttClientOptions options)
    {
            EventLog Log1 = new EventLog();

            if (!EventLog.SourceExists("IsMQTTClient"))
                EventLog.CreateEventSource("IsMQTTClient", "IsMQTTClient");

            Log1.Source = "IsMQTTClient";
            Log1.Log = "IsMQTTClient";


            if (modalita.ToString().ToUpper() == "SQL")
        {
            SqlConnection conn;

                foreach (var database in GlobalVar.SQL_SERVER_DATABASE_NAME)
            {
                conn = new SqlConnection("Server=" + GlobalVar.SQL_SERVER_NAME + ";Database=" + database.Trim() + ";User Id=" + GlobalVar.SQL_SERVER_USER + ";Password=" + GlobalVar.SQL_SERVER_PASSWORD + ";");

                // inserisco maccd_log_analitico
                var QUERY = new SqlCommand(@"SELECT * FROM [001].[IS_MQTT_TOPIC_SUBSCRIPTION] WHERE ENABLE = 1", conn);

                conn.Open();

                try
                {
                    DataTable dt = new DataTable();
                    dt.Load(QUERY.ExecuteReader());
                    conn.Close();

                        foreach(DataRow row in dt.Rows)
                        {
                            if (row["QOS"].ToString() == "2")
                            {
                                await mqttClient.SubscribeAsync(new MqttTopicFilterBuilder().WithTopic(row["TOPIC"].ToString()).WithExactlyOnceQoS().Build());
                                Log1.WriteEntry("### SUBSCRIBED TO TOPIC: " + row["TOPIC"].ToString() + " QoS: 2 ###");
                            }
                            else if (row["QOS"].ToString() == "1")
                            {
                                await mqttClient.SubscribeAsync(new MqttTopicFilterBuilder().WithTopic(row["TOPIC"].ToString()).WithAtLeastOnceQoS().Build());
                                Log1.WriteEntry("### SUBSCRIBED TO TOPIC: " + row["TOPIC"].ToString() + " QoS: 1 ###");
                            }
                            else
                            {
                                await mqttClient.SubscribeAsync(new MqttTopicFilterBuilder().WithTopic(row["TOPIC"].ToString()).WithAtMostOnceQoS().Build());
                                Log1.WriteEntry("### SUBSCRIBED TO TOPIC: " + row["TOPIC"].ToString() + " Qos: 0 ###");
                            }
                        }
                }
                catch (Exception ex)
                {
                        Log1.WriteEntry(ex.ToString());
                    conn.Close();
                }
            }
        }
        else
        {
            await mqttClient.SubscribeAsync(new MqttTopicFilterBuilder()
                                            .WithTopic("#")
                                            .WithAtLeastOnceQoS() // <-- This sets the QoS of the subscription
                                            .Build());

                Log1.WriteEntry("### SUBSCRIBED TO ALL TOPICS ###");
        }

    }
    }`

and the field "modalita" is passed with value => "ALL", so it should subscribe to all topics with QoS 1

Any other tips?

**Comment by davemcdermid** [comment]
Sorry, it's hard to follow as there's lots of code here not related to this library, and depending on state.

If you want the broker to persist your session, make sure you:

* Connect without a clean session `WithCleanSession(false)`
* Use a consistent ClientId `.WithClientId("myClientId")`
* Subscribe with QoS >=1 (see above)
* Disconnect subsciber
* Publish messages with QoS >= 1
* Reconnect subscriber same as first time

These together worked for me.

**Comment by swIsoftSrl** [comment]
I think it's the procedure i'm following
Are you using v3.0.14?

it seems like i'm doing the exact procedure you described above

The options for my server are:

`.WithClientId("IsoftMQTTBroker")
                .WithConnectionValidator()//code for connection validation insede
                .WithConnectionBacklog(100)
                .WithDefaultEndpointPort(Int32.Parse((GlobalVar.MQTT_BROKER_PORT)))
                .WithPersistentSessions();`


The options for my client are:

` options = (MqttClientOptions)new MqttClientOptionsBuilder()
                        .WithClientId("IsoftMqttClientWS")
                        .WithTcpServer(broker ip, broker port) // Port is optional
                        .WithCredentials(client username, client password)
                        .WithSessionExpiryInterval(0)
                        .WithCommunicationTimeout(new TimeSpan(24, 0, 0))
                        .WithCleanSession(false)
                        .Build();`


The client each reconnect subscribes with
`await mqttClient.SubscribeAsync(new MqttTopicFilterBuilder()
                                            .WithTopic("#")
                                            .WithAtLeastOnceQoS() // <-- This sets the QoS of the subscription
                                            .Build());`
I tested also with ".WithExactlyOnceQoS()"

An example message sent while Subscriber is offline was:

### RECEIVED APPLICATION MESSAGE ### 12/02/2021 16:05:54.734
+ Topic = test2
+ Payload = test2
+ QoS = ExactlyOnce
+ Retain = False

Logged like that on broker side

No luck on getting it back on subscriber reconnection

Thanks

**Comment by davemcdermid** [comment]
Sorry I can't help further, I'm really not an expert here. I found http://mqtt-explorer.com/ useful for visualising what was happening, and Wireshark for digging into what was really being sent & received. Not sure if that's helpful for you.

**Comment by swIsoftSrl** [comment]
Hi @SeppPenner ,
you mean that using managed client could solve this issue or what?
Thanks

**Comment by SeppPenner** [comment]
No, I guess, I somehow linked the wrong issue... I don't think that the managed client would help here.

**Comment by dkavicki** [comment]
fwiw I had the same issue until I did .WithSessionExpiryInterval(9999)
Then it worked for me. (with the ManagedClient) 
However, I've noticed that my test Mqttnet server would send the message every time (and not just once) whenever I start and connect the client again
